### PR TITLE
Speed up beatmap re-imports

### LINKED ISSUE #12746: Investigate beatmap import performance when beatmap is already loaded into client
While this is marked as a feature request, I'm really just looking for a quick hack because I'm not sure how applicable this issue is to the userbase at large. So it's more of just a question. Please feel free to close it immediately.

# Background

I have a very large (27,300) collection of beatmaps that I'm trying to import from stable. As you can imagine, not all of these are well-formed beatmaps that lazer will accept. It's also not the most stable process in the world; my game crashed last time I tried to load the selection screen, forcing me to start the import over even though about 20k beatmaps are now successfully imported.

# Speculative implementation

Without any familiarity with the codebase, I was thinking of just adding a new table to the database that contains a unique index of directory names within `osu!\Songs`. Then all the import process has to do is look at the directory name before determining whether to skip, so it can get back to where it left off last time as quickly as possible. Once a beatmap is successfully (or unsuccessfully, I suppose) added to the collection, it adds the stable directory name.

# Questions

Is my approach too naive? Do you have any suggestions? Would this feature actually be desirable for a pull request?

Thank you!

### PR DESCRIPTION
**Pull request by smoogipoo**
Resolves https://github.com/ppy/osu/issues/12746

Reimports my entire collection of 2000 beatmaps in 3 seconds. Doesn't handle the case of un-deleting existing non-purged beatmaps, but I see that as an exception rather than the rule.

### CODE REVIEW

File: osu.Game/Beatmaps/BeatmapManager.cs [hunk 1]
```diff
@@ -403,6 +403,32 @@ protected override BeatmapSetInfo CreateModel(ArchiveReader reader)
                 return null;
             }
 
+            // pre-check to make sure that the archive hasn't already been imported.
+            bool anyToImport = false;
+
+            foreach (var file in reader.Filenames.Where(f => f.EndsWith(".osu", StringComparison.OrdinalIgnoreCase)))
```
Comment (smoogipoo) [inline osu.Game/Beatmaps/BeatmapManager.cs:409]: Note that this doesn't consider non `.osu` files in the archive. I don't think this a big deal, but checking all files reduces performance substantially even though it's still much faster than current master.

**Comment by peppy** [comment]
Which piece of potentially important import process is this skipping? The existing process already does hash comparisons of only .osu files, so I'm not sure why this is being added a second time (without any documentation of why it's required / which portion is causing the overhead).

I think rather than hacking around the existing process, it should be fixed at an `ArchiveModelManager` level so this also supports other managers. Especially as we are looking towards a refactor with realm in mind (where performance gains were already seen in the import process).

If we are going to consider this direction, this needs an inline comment mentioning it's a temporary workaround and not a permanent fixture.

Also important to keep in mind this breaks the previously established behaviour of allowing reimporting to fix broken files. Now, as long as the .osu files are in the file store, regardless of whether the rest of the files (or even the beatmap/beatmapset) are there, reimporting to fix issues will be impossible. I can't see this as a good direction.

**Comment by smoogipoo** [comment]
I think of most importance, this skips:
- Reading beatmaps
- Calling the difficulty calculator

Both of which are done in the `Populate()` stage so as to be outside of the write context.

Of secondary importance, this also skips hashing all files in the archive, though I know you're going to disagree with this. No matter what we do, this isn't going to get any faster.
So what, should we just not have this? This is a personal blocker for me, because I'm not ready to switch to lazer 100%.

Honestly, why are you reimporting _every beatmap_ to fix beatmaps? Find the broken one, delete it, and then re-import it. As I mentioned in the OP, this doesn't handle/undelete deleted beatmaps.
And if you're reimport _every_ beatmap to fix beatmaps, this PR brings the whole reimport process to literally a few seconds as opposed to 10+ minutes. It'll find the deleted beatmap, reimport it, then complete the rest of the process in the next 2 seconds.

**Comment by peppy** [comment]
Yeah, I think we need to restructure the `Populate` call order to avoid that from happening (I know there's a reason it's done in that order, but don't believe it can't be worked around).

As for your second point, a special mode for the "import from stable" flow (ie. a "quick" flag) may be a good path forward. Just strongly against duplicating the existing hash comparison that is already part of the standard flow.

**Comment by smoogipoo** [comment]
There's quite some pushback against this change so I'm pulling it for now. Fixing tests requires considerable work that I want to have a clear path forward before doing anything more.

**Pull request closed without merging**